  **Regressão Linear**

Exercício 1

---
Parte 1- 

Abra o banco de dados Boston [(conhecer melhor sobre esse dataset? click aqui:)](https://www.cs.toronto.edu/~delve/data/boston/bostonDetail.html)e estime 3 regressões:

1.   Prevendo medv usando como preditor lstat;
2.   Prevendo medv usando como preditor age;
3.   Usando ambos os preditores.




In [0]:
import pandas as pd # fornece ferramentas de análise de dados
import statsmodels.formula.api as sms # fornece classes e funções para a estimativa de modelos estatísticos
import matplotlib.pyplot as plt  # fornece uma maneira de plotagem semelhante ao MATLAB.
import numpy as np # fornece ferramentas para realizar cálculos em Arrays Multidimensionais
# %matplotlib inline (a saída dos comandos de plotagem é exibida em linha nos frontends)
%matplotlib inline 

In [3]:
# Lendo o dataset boston 
boston = pd.DataFrame.from_csv("https://raw.githubusercontent.com/selva86/datasets/master/BostonHousing.csv")
boston.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: from_csv is deprecated. Please use read_csv(...) instead. Note that some of the default arguments are different, so please refer to the documentation for from_csv when changing your function calls
  """Entry point for launching an IPython kernel.


zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
crim                                                                        
0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

              b  lstat  medv  
crim                          
0.00632  396.90   4.98  24.0  
0.02731  396.90   9.14  21.6  
0.02729  392.83   4.03  34.7  
0.03237  394.63   2.94  33.4  
0.06905  396.90   5.33  36.2

In [0]:
# Primeira estimativa: medv usando como preditor lstat

reg1 = sms.ols(formula="medv ~ lstat", data=boston).fit()
print (reg1.summary())
 


In [0]:
# Segunda estimativa: medv usando como preditor age

reg2 = sms.ols(formula="medv ~ age", data=boston).fit()
print (reg2.summary())

In [0]:
# Terceira estimativa: usando ambos como preditor

reg3 = sms.ols(formula= "medv ~ lstat + age", data=boston).fit()
print(reg3.summary())

O que podemos dizer sobre o comportamento dos coeficientes estimados?
E sobre o ajuste dos modelos?



Parte 2 -  


Estime uma regressão usando todas as variáveis como preditoras. Todas são úteis para prever medv? Quais você deixaria e quais tiraria? Por quê?

*   Como foi o ajuste desse modelo?

*   Agora tire as variáveis que você considerou que não ajudam a estimação de medv. O que aconteceu com o ajuste do modelo?





In [0]:
reg4 = sms.ols(formula= "medv ~ zn + indus + chas + nox + rm + age \
               + dis + rad + tax + ptratio + b + lstat", data = boston).fit()
print(reg4.summary())

feedback da análise >here<

Exercício 2


---



Vamos fazer algumas investigações no banco credit. Aí estão dados de cartão de crédito com as seguintes variáveis:

Income - renda em milhares de dólares;
Limit - limite de crédito;
Rating - o score de crédito;
Card - número de cartões de crédito que a pessoa possui;
Age - idade;
Education - anos de escolaridade;
Gender - sexo;
Student - se é estudante;
Married - se é casado;
Ethnicity - etnia e
Balance - dívida média de cartão de crédito.


---


In [29]:

import numpy as np
import pandas as pd
import statsmodels.formula.api as sms

# Lendo o dataset credit
credit = pd.read_csv("http://www-bcf.usc.edu/~gareth/ISL/Credit.csv")
# Tira a primeira coluna
print(credit.columns)
credit = credit.drop(columns=['Unnamed: 0'])
# Após retirar coluna com índices...
print(credit.columns)
credit.head()


Index(['Unnamed: 0', 'Income', 'Limit', 'Rating', 'Cards', 'Age', 'Education',
       'Gender', 'Student', 'Married', 'Ethnicity', 'Balance'],
      dtype='object')
Index(['Income', 'Limit', 'Rating', 'Cards', 'Age', 'Education', 'Gender',
       'Student', 'Married', 'Ethnicity', 'Balance'],
      dtype='object')


Income  Limit  Rating  Cards  Age  Education  Gender Student Married  \
0   14.891   3606     283      2   34         11    Male      No     Yes   
1  106.025   6645     483      3   82         15  Female     Yes     Yes   
2  104.593   7075     514      4   71         11    Male      No      No   
3  148.924   9504     681      3   36         11  Female      No      No   
4   55.882   4897     357      2   68         16    Male      No     Yes   

   Ethnicity  Balance  
0  Caucasian      333  
1      Asian      903  
2      Asian      580  
3      Asian      964  
4  Caucasian      331

1-
Estude a diferença entre o*** balance ***entre*** homens e mulheres ***usando um modelo de regressão. Interprete os resultados.



In [21]:
reg1 = sms.ols("Balance ~ Gender", data=credit).fit()
print(reg1.summary())

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.1836
Date:                Mon, 18 Mar 2019   Prob (F-statistic):              0.669
Time:                        19:08:06   Log-Likelihood:                -3019.3
No. Observations:                 400   AIC:                             6043.
Df Residuals:                     398   BIC:                             6051.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        529.5362     31.988     16.

O P-value preditivo de Gender/sexo é 0.669. Sendo então maior que 0.05 e consequentemente o efeito é considerado nulo.



---


2-
Estude a diferença entre o ***balance*** entre pessoas de diferentes etnias usando um modelo de regressão. Interprete os resultados.


In [23]:
reg2 = sms.ols("Balance ~ Ethnicity", data=credit).fit()
print(reg2.summary())

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.005
Method:                 Least Squares   F-statistic:                   0.04344
Date:                Mon, 18 Mar 2019   Prob (F-statistic):              0.957
Time:                        19:08:44   Log-Likelihood:                -3019.3
No. Observations:                 400   AIC:                             6045.
Df Residuals:                     397   BIC:                             6057.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept                531

O P-value preditivo de Ethnicity é 0.774 (Asian) e 0.826 (Caucasian). Sendo então maior que 0.05 e consequentemente o efeito é considerado nulo.

---

3-
Estime agora um modelo de regressão utilizando todas as variáveis presentes. O que podemos dizer sobre os resultados? Todas as variáveis ajudam a explicar o balance? Quais variáveis você tiraria? Como ficaria seu novo modelo?

In [26]:
reg3 = sms.ols("Balance ~ Income + Limit + Rating + Cards + Age + Education \
               + Gender + Student + Married + Ethnicity", data=credit).fit()
print(reg3.summary())

                            OLS Regression Results                            
Dep. Variable:                Balance   R-squared:                       0.955
Model:                            OLS   Adj. R-squared:                  0.954
Method:                 Least Squares   F-statistic:                     750.3
Date:                Mon, 18 Mar 2019   Prob (F-statistic):          1.11e-253
Time:                        19:26:37   Log-Likelihood:                -2398.7
No. Observations:                 400   AIC:                             4821.
Df Residuals:                     388   BIC:                             4869.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept               -489